### ISSUE TITLE
Ambiguous behaviour of tf.train.global_step

### ISSUE THREAD

**Issue by kingspp** [tensorflow/tensorflow#26085]
**System information**
- OS Platform and Distribution: *Darwin - Mac OS X 10.14*
- TensorFlow installed from (source or binary): *Source*
- TensorFlow version (use command below): *v1.12.0-rc2-3-ga6d8ffae09 1.12.0*
- Python version: *3.6.5*

**Describe the current behavior**

Let me illustrate the ambiguous behaviour of global step when using iterators and custom gradient operations. For the example, I will consider a toy problem, Input is a list of 10 digits filled with 1. Output is a list of 10 digits filled with 1-10. A variable is added to the input, so during the course of training, Variable mimics the operation and finally predicts a list of 10 digits filled with 1-10.

**Assumption: (Pls correct if wrong)**
1. Global step is created using `tf.train.get_or_create_global_step()` and is fed to global_step variable in optimizer
2. Global step increments for every run of optimizer.
3. Global step is fetched using `tf.train.get_global_step()` for display purpose.

#### 1. Basic Optimization using Initializable iterators as Data Feed. (Expected)
```python
import tensorflow as tf

def gen():
    for i in range(10):
        yield [1,1,1,1,1,1,1,1,1,1]

itr = tf.data.Dataset.from_generator(generator=gen, output_types=tf.float32)
ini = itr.make_initializable_iterator()

pred = ini.get_next() +  tf.Variable(initial_value=tf.random_normal(shape=(1,10)), dtype=tf.float32)
label = tf.constant([1,2,3,4,5,6,7,8,9,10], shape=(1,10), dtype=tf.float32)
loss = tf.reduce_mean(tf.square(pred - label))
training_op = tf.train.AdamOptimizer(learning_rate=0.1).minimize(loss=loss, global_step=tf.train.get_or_create_global_step())

with tf.Session() as sess:
    sess.run(tf.group([tf.global_variables_initializer(), tf.local_variables_initializer(), tf.tables_initializer()]))
    for i in range(10):
        sess.run(ini.initializer)
        print('Global Step: ', sess.run(tf.train.get_global_step()))
        try:
            while True:
                _,l, g = sess.run([training_op, loss, tf.train.get_or_create_global_step()])
        except tf.errors.OutOfRangeError:
            pass
"""
Output - Expected:
Global Step:  0
Global Step:  10
Global Step:  20
Global Step:  30
Global Step:  40
Global Step:  50
Global Step:  60
Global Step:  70
Global Step:  80
Global Step:  90
"""
``` 

#### 2. Replace `Optimizer.minimize` with `Optimizer.compute_gradients` and `Optimizer.apply_gradients` with basic gradient manipulation. (Expected)
```python
import tensorflow as tf

def gen():
    for i in range(10):
        yield [1,1,1,1,1,1,1,1,1,1]

itr = tf.data.Dataset.from_generator(generator=gen, output_types=tf.float32)
ini = itr.make_initializable_iterator()

pred = ini.get_next() +  tf.Variable(initial_value=tf.random_normal(shape=(1,10)), dtype=tf.float32)
label = tf.constant([1,2,3,4,5,6,7,8,9,10], shape=(1,10), dtype=tf.float32)
loss = tf.reduce_mean(tf.square(pred - label))
opt = tf.train.AdamOptimizer(learning_rate=0.1)
compute_op = opt.compute_gradients(loss=loss)
# Manipulate existing gradients
xyz = [(tf.nn.sigmoid(i[0]), i[1]) for i in compute_op]
training_op = opt.apply_gradients(grads_and_vars=xyz, global_step=tf.train.get_or_create_global_step())

with tf.Session() as sess:
    sess.run(tf.group([tf.global_variables_initializer(), tf.local_variables_initializer(), tf.tables_initializer()]))
    for i in range(10):
        sess.run(ini.initializer)
        print('Global Step: ', sess.run(tf.train.get_global_step()))
        try:
            while True:
                _,l, g = sess.run([training_op, loss, tf.train.get_or_create_global_step()])
        except tf.errors.OutOfRangeError:
            pass
"""
Output - Expected:
Global Step:  0
Global Step:  10
Global Step:  20
Global Step:  30
Global Step:  40
Global Step:  50
Global Step:  60
Global Step:  70
Global Step:  80
Global Step:  90
"""
```

#### 3. Similar to Step 2, but instead of operating on the gradient tensor, Gradients are replaced with `tf.zeros()` (Ambiguous)
```python
import tensorflow as tf

def gen():
    for i in range(10):
        yield [1,1,1,1,1,1,1,1,1,1]

itr = tf.data.Dataset.from_generator(generator=gen, output_types=tf.float32)
ini = itr.make_initializable_iterator()

pred = ini.get_next() +  tf.Variable(initial_value=tf.random_normal(shape=(1,10)), dtype=tf.float32)
label = tf.constant([1,2,3,4,5,6,7,8,9,10], shape=(1,10), dtype=tf.float32)
loss = tf.reduce_mean(tf.square(pred - label))
opt = tf.train.AdamOptimizer(learning_rate=0.1)
compute_op = opt.compute_gradients(loss=loss)
# Replace Gradients
xyz = [(tf.zeros(shape=i[0].shape.as_list()), i[1]) for i in compute_op]
training_op = opt.apply_gradients(grads_and_vars=xyz, global_step=tf.train.get_or_create_global_step())

with tf.Session() as sess:
    sess.run(tf.group([tf.global_variables_initializer(), tf.local_variables_initializer(), tf.tables_initializer()]))
    for i in range(10):
        sess.run(ini.initializer)
        print('Global Step: ', sess.run(tf.train.get_global_step()))
        try:
            while True:
                _,l, g = sess.run([training_op, loss, tf.train.get_or_create_global_step()])
        except tf.errors.OutOfRangeError:
            pass
"""
Output - Ambiguous:
Global Step:  0
Global Step:  11
Global Step:  22
Global Step:  33
Global Step:  44
Global Step:  55
Global Step:  66
Global Step:  77
Global Step:  88
Global Step:  99
"""
```

**Describe the expected behavior**
#### Expected behaviour for step 3
```python
"""
Output - Expected:
Global Step:  0
Global Step:  10
Global Step:  20
Global Step:  30
Global Step:  40
Global Step:  50
Global Step:  60
Global Step:  70
Global Step:  80
Global Step:  90
"""
```

1. Why / How does global_step gets an extra increment in Step 3?
2. Is it a bad idea to replace gradient tensors with custom tensors (Probably custom tensors do not obey tf.GraphKeys.UpdateOps)?
3. What is the right method to replace gradients and proposed validation?

**Comment by alextp** [comment]
The reason why you see the global step numbers out of sync is that there is no control dependency between the read of the variable in get_or_create_global_step and the increment to it in session.run. Because they are unordered tensorflow can either give you the old value or the new value; the tf2 implementastion of variables prefers to give you the value before update while the tf1 prefers to give you the value after.

If you want to know which value you're getting you should add control dependencies to decide what happens. Note that this is only an issue for the global step because it's invisibly created and invisibly modified; this API endpoint (get_or_create_global_step) has been removed in tf2 because of this type of issue.

So, short answer is that this is WAI.
